### PR TITLE
[MPS] Fix bidirectional LSTM & small one-direction LSTM fix

### LINKED ISSUE #94754: BiLSTM incorrect on MPS
### 🐛 Describe the bug

Hi, I been testing various NLP models on MPS, and I observe various strange things. It seems that bidirectional LSTM on MPS is not working correctly.

```
import torch
import torch.nn as nn

print("Torch version: ", torch.__version__)
torch.manual_seed(123)
embeddings = torch.rand(3,1,2)  # L, B, E

model = nn.LSTM(input_size=2, hidden_size=2, bidirectional=True)
print(model(embeddings)[0])

model.to("mps")
print(model(embeddings.to("mps"))[0])
```

Results are below. CPU above, MPS below. How can results on CPU be different than on MPS?

```
Torch version:  2.0.0.dev20230205
tensor([[[ 0.0054,  0.2338, -0.1167,  0.2263]],
        [[ 0.0200,  0.3305, -0.0967,  0.2075]],
        [[ 0.0422,  0.3740, -0.0590,  0.1553]]], grad_fn=<CatBackward0>)
tensor([[[0.2325, 0.2447, 0.4149, 0.4394]],
        [[0.3524, 0.3702, 0.3516, 0.3785]],
        [[0.4139, 0.4346, 0.2306, 0.2531]]], device='mps:0',
       grad_fn=<CatBackward0>)
```

### Versions

I’m testing on MacBook Pro with M1PRO, 32GB, 10-core CPU, 16-core GPU.

```
PyTorch version: 2.0.0.dev20230205
Is debug build: False
CUDA used to build PyTorch: None
ROCM used to build PyTorch: N/A

OS: macOS 13.1 (x86_64)
GCC version: Could not collect
Clang version: 14.0.0 (clang-1400.0.29.202)
CMake version: Could not collect
Libc version: N/A

Python version: 3.9.16 (main, Jan 11 2023, 10:02:19)  [Clang 14.0.6 ] (64-bit runtime)
Python platform: macOS-10.16-x86_64-i386-64bit
Is CUDA available: False
CUDA runtime version: No CUDA
CUDA_MODULE_LOADING set to: N/A
GPU models and configuration: No CUDA
Nvidia driver version: No CUDA
cuDNN version: No CUDA
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Apple M1 Pro

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] torch==2.0.0.dev20230205
[pip3] torchaudio==2.0.0.dev20230205
[pip3] torchvision==0.15.0.dev20230205
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  h0a44026_0    pytorch
[conda] mkl                       2021.4.0           hecd8cb5_637  
[conda] mkl-service               2.4.0            py39h9ed2024_0  
[conda] mkl_fft                   1.3.1            py39h4ab4a9b_0  
[conda] mkl_random                1.2.2            py39hb2f4e1b_0  
[conda] numpy                     1.23.5           py39he696674_0  
[conda] numpy-base                1.23.5           py39h9cd3388_0  
[conda] pytorch                   2.0.0.dev20230205         py3.9_0    pytorch-nightly
[conda] torchaudio                2.0.0.dev20230205        py39_cpu    pytorch-nightly
[conda] torchvision               0.15.0.dev20230205        py39_cpu    pytorch-nightly
```

cc @kulinseth @albanD @malfet @DenisVieriu97 @razarmehr @abhudev

### PR DESCRIPTION
**Pull request by alexdremov**
Fixes #94754

With this PR I hope to finish my breathtaking journey of fixing MPS LSTM.

Here, I enable `bidirectional` on MPS. Also, I've noticed that cache key did not account for all parameters, so there could have been problems with one-directional LSTM when created without bias or dropout and then with one of them.

cc: @kulinseth @jhavukainen

### CODE REVIEW

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (AlexRoar) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:38]: This substantial part of code is duplicated in forward and backward, so incapsulated it in function

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (AlexRoar) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:146]: Here, bug appeared when LSTM was created without biases and then with biases. Added test case

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (AlexRoar) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:292]: Turns out, I left it unused in the previous PR

File: test/test_mps.py
Comment (AlexRoar) [inline test/test_mps.py:9127]: Added `hx` and `cx` to function to test how gradients flow through them

File: test/test_mps.py
Comment (AlexRoar) [inline test/test_mps.py:9127]: Added hx and cx to function to test how gradients flow through them

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (jhavukainen) [reply aten/src/ATen/native/mps/operations/RnnOps.mm:146]: Good catch

**Review by jhavukainen — APPROVED** [review]
All looks good to me, well done. Nice catch on the incorrect caching behavior with the biases.

**Comment by kulinseth** [comment]
Looks great.

**Review by kulinseth — APPROVED** [review]

**Comment by AlexRoar** [comment]
@kulinseth merge failed(

**Comment by AlexRoar** [comment]
Just rebased

**Comment by kulinseth** [comment]
@albanD , @malfet , @soulitzer , can one of you please take a quick look at this ? MPS changes look good. Need approval from Core maintainers on this.

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:753]: Nit
```suggestion
            if (has_biases) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:740]:
```suggestion
        for (const auto  i : c10::irange(num_layers * (bidirectional ? 2 : 1))) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:702]:
```suggestion
        for (const auto I : c10::irange(num_layers * (bidirectional ? 2 : 1))) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:410]:
```suggestion
                    for (const auto i : c10::irange(num_layers * (bidirectional ? 2 : 1))) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:120]:
```suggestion
    for (const auto i : c10::irange(num_layers * (bidirectional ? 2 : 1))) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:364]:
```suggestion
    for (const auto i : c10::irange(num_layers * (bidirectional ? 2 : 1)) {
```

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (malfet) [inline aten/src/ATen/native/mps/operations/RnnOps.mm:756]: Not super familiar with the code, would be good to leave a comment why `grad_bias` should be pushed twice.

**Review by malfet — APPROVED** [review]
LGTM, few general remarks:
- Please use `for(const auto i : c10::irange(limit))` to iterate over list of elements
- `num_layer * (bidirectional ? 2 :1 )` is used frequently in the code, perhaps a helper variable , say `total_num_layer` can be defined at the start of the routine and then user throughout the codebase?

File: aten/src/ATen/native/mps/operations/RnnOps.mm
Comment (AlexRoar) [reply aten/src/ATen/native/mps/operations/RnnOps.mm:756]: Sure. I've added several comments near complicated pieces. Will extend them in commit with your fixes

Comment (AlexRoar) [reply aten/src/ATen/native/mps/operations/RnnOps.mm:768]: Added explanation

**Pull request closed without merging**
